keyboards/acheron/shark/beta/keymaps/default/keymap.c

```diff
@@ -0,0 +1,52 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_ortho_4x12(
+        KC_TAB  , KC_Q    , KC_W    , KC_E    , KC_R    , KC_T    , KC_Y    , KC_U    , KC_I    , KC_O    , KC_P    , KC_BSPC,
+        KC_ESC  , KC_A    , KC_S    , KC_D    , KC_F    , KC_G    , KC_H    , KC_J    , KC_K    , KC_L    , KC_SCLN , KC_QUOT,
+        KC_LSFT , KC_Z    , KC_X    , KC_C    , KC_V    , KC_B    , KC_N    , KC_M    , KC_COMM , KC_DOT  , KC_SLSH , KC_ENT ,
+        KC_LCTL , KC_LALT , KC_LGUI , KC_NO   , MO(1)   , KC_SPC  , KC_SPC  , MO(2)   , KC_LEFT , KC_DOWN , KC_UP   , KC_RGHT
+    ),
+    [1] = LAYOUT_ortho_4x12(
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , 
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS 
+    ),
+    [2] = LAYOUT_ortho_4x12(
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , 
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS 
+    ),
+    [3] = LAYOUT_ortho_4x12(
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , 
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS ,
+        KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS 
+    )
+};
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+        if (clockwise) {
+            tap_code_delay(KC_VOLU, 10);
+        } else {
+            tap_code_delay(KC_VOLD, 10);
+        }
+    return false;
+}
```

Comment:
Could you move this to the keyboard's C file, and using the kb function instead?

Reply:
The idea here is that the encoder callback be user-adjustable; for instance, the `gondolindrim` layout makes use of a heavily customized `encoder_update_user`. People have also reported wanting to tune their encoder callbacks on the revision Alpha of Shark, hence the function here

Reply:
Yeah, I do understand that, but: 

This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```c
#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

This allows for it to be user customizable *and* to set a default behavior.

Reply:
I see. Not to lessen your implementation, but wouldn't that make it so both pieces of code make it into the firmware?

What about using a new macro `USER_ENCODER`, something like this:

On `beta.c`:

```c
#if defined(ENCODER_ENABLE) && !defined(USER_ENCODER)
bool encoder_update_kb(uint8_t index, bool clockwise) {
        if (index == 0) {
                if (clockwise) tap_code_delay(KC_VOLU, 10);
                else tap_code_delay(KC_VOLD, 10);
        }   
        return true;
}
#endif
```

On `keymap.c`:
```c
#ifdef ENCODER_ENABLE
#define USER_ENCODER
bool encoder_update_kb(uint8_t index, bool clockwise) {
    <whatever the user wants this to do...>
}
#endif
```

Reply:
Well, returning false would prevent the kb level function from calling the rest of the code.  And from what I understand, from there it depends on the compiler if it drops the code. 

But yeah, the user function, you'd want to return false, using this.

Reply:
Well I tried doing it and ideed, if I define `USER_ENCODER` in the keymap the compiler will act like it was not defined, hence `encoder_update_kb` will be compiled like it was not defined, probably because the keymap is compiled after the `keyboard.c` although I though macros were processed before compile time. It is what it is.

I think the best option here was to use this in `config.h` which does what we want, please check 2ed915e088 . If you think that's fine I'll add a note on README and we should be done